### PR TITLE
445 readunread jobchat notifications - front end

### LINKED ISSUE #445: Read/Unread - JobChat Notifications - Front end
As a chat user
I want to be able to be notified of new posts to chats and record whether I have read a chat or not
so that I can be aware of new messages coming and so that I can see which chats I have looked at, or still need to look at.

### PR DESCRIPTION
**Pull request by camerojo**
A lot of changes here - but mainly they involved moving towards a common code base for the chat components across admin portal and candidate portal. Didn't get there 100% but much closer.

The key files to look at are

chat.service and the new chat-read-status component

That is where all the heavy lifting for front end notifications sits.
Also files that call the chatService.getOrCreate method which returns an observable to a chat from a chat request.

### CODE REVIEW

File: ui/admin-portal/src/app/app.module.ts [hunk 1]
```diff
@@ -843,6 +844,8 @@ import {
     SearchHomeComponent,
     JobHomeComponent,
     ListHomeComponent,
+    ChatReadStatusComponent,
+    ListHomeComponent,
```
Comment (sadatmalik) [inline ui/admin-portal/src/app/app.module.ts:848]:
```suggestion

```

Duplicated 2 lines above - safe to delete this

File: ui/admin-portal/src/app/components/chat/chat-read-status/chat-read-status.component.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+  }
+
+  ngOnDestroy(): void {
+    this.unsubscribe();
+  }
+
+  private subscribeForChatUpdates() {
+    if (this.chats) {
+      this.unsubscribe();
```
Comment (sadatmalik) [inline ui/admin-portal/src/app/components/chat/chat-read-status/chat-read-status.component.ts:47]:
```suggestion
      this.unsubscribe();
      this.unreadIndicator = ''
```

Will handle the case where there are no chats yet

File: ui/admin-portal/src/app/components/chat/chat-read-status/chat-read-status.component.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      this.setIndicator(allChatsRead);
+      this.subscribeForChatUpdates();
+    }
+  }
+
+  ngOnDestroy(): void {
+    this.unsubscribe();
+  }
+
+  private subscribeForChatUpdates() {
```
Comment (sadatmalik) [inline ui/admin-portal/src/app/components/chat/chat-read-status/chat-read-status.component.ts:45]: Neat implementation - love this!

**Review by sadatmalik — APPROVED** [review]
Approved

File: ui/admin-portal/src/app/app.module.ts (hunk 1 above)
Comment (camerojo) [reply ui/admin-portal/src/app/app.module.ts:848]: Done - thanks

**Pull request merged**